File: ios/RNAdMobBannerViewManager.m
#import "RNAdMobBannerView.h"

@interface RNAdMobBannerViewManager : RCTViewManager
@end

@implementation RNAdMobBannerViewManager

RCT_EXPORT_MODULE();

- (UIView *)view
{
    return [RNAdMobBannerView new];
}

RCT_EXPORT_METHOD(requestAd:(nonnull NSNumber *)reactTag)
{
    [self.bridge.uiManager addUIBlock:^(__unused RCTUIManager *uiManager, NSDictionary<NSNumber *, RNAdMobBannerView *> *viewRegistry) {
        RNAdMobBannerView *view = viewRegistry[reactTag];
        if (![view isKindOfClass:[RNAdMobBannerView class]]) {
            RCTLogError(@"Invalid view returned from registry, expecting RNAdMobBannerView, got: %@", view);
        } else {
            [view requestAd];
        }
    }];
}

RCT_EXPORT_VIEW_PROPERTY(size, NSString)
RCT_EXPORT_VIEW_PROPERTY(unitId, NSString)
RCT_EXPORT_VIEW_PROPERTY(sizes, NSArray)
RCT_EXPORT_VIEW_PROPERTY(requestOptions, NSDictionary)

RCT_EXPORT_VIEW_PROPERTY(onSizeChange, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onAdLoaded, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onAdFailedToLoad, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onAdOpened, RCTDirectEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onAdClosed, RCTDirectEventBlock)

RCT_EXPORT_VIEW_PROPERTY(onAppEvent, RCTDirectEventBlock)

@end

File: android/src/main/java/com/rnadmob/admob/ads/banner/RNAdMobBannerViewManager.java
package com.rnadmob.admob.ads.banner;

import androidx.annotation.Nullable;

import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.common.MapBuilder;
import com.facebook.react.uimanager.SimpleViewManager;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.annotations.ReactProp;

import java.util.Map;

import javax.annotation.Nonnull;

public class RNAdMobBannerViewManager extends SimpleViewManager<RNAdMobBannerView> {
    public static final String REACT_CLASS = "RNAdMobBannerView";
    public static final String EVENT_AD_LOADED = "onAdLoaded";
    public static final String EVENT_AD_FAILED_TO_LOAD = "onAdFailedToLoad";
    public static final String EVENT_AD_OPENED = "onAdOpened";
    public static final String EVENT_AD_CLOSED = "onAdClosed";
    public static final String EVENT_SIZE_CHANGE = "onSizeChange";
    public static final String EVENT_APP_EVENT = "onAppEvent";

    private static final String COMMAND_REQUEST_AD = "requestAd";

    @Nonnull
    @Override
    public String getName() {
        return REACT_CLASS;
    }

    @Nonnull
    @Override
    public RNAdMobBannerView createViewInstance(@Nonnull ThemedReactContext themedReactContext) {
        return new RNAdMobBannerView(themedReactContext);
    }

    @Override
    @Nullable
    public Map<String, Object> getExportedCustomDirectEventTypeConstants() {
        MapBuilder.Builder<String, Object> builder = MapBuilder.builder();
        String[] events = {
                EVENT_SIZE_CHANGE,
                EVENT_AD_LOADED,
                EVENT_AD_FAILED_TO_LOAD,
                EVENT_AD_OPENED,
                EVENT_AD_CLOSED
        };
        for (String event : events) {
            builder.put(event, MapBuilder.of("registrationName", event));
        }
        return builder.build();
    }

    @ReactProp(name = "unitId")
    public void setUnitId(RNAdMobBannerView view, String unitId) {
        view.setUnitId(unitId);
    }

    @ReactProp(name = "size")
    public void setSize(RNAdMobBannerView view, String size) {
        view.setSize(size);
    }

    @ReactProp(name = "sizes")
    public void setSizes(RNAdMobBannerView view, ReadableArray adSizeStrings) {
        view.setSizes(adSizeStrings);
    }

    @ReactProp(name = "requestOptions")
    public void setRequestOptions(RNAdMobBannerView view, ReadableMap requestOptions) {
        view.setRequestOptions(requestOptions);
    }

    @Override
    public void receiveCommand(@Nonnull RNAdMobBannerView bannerView, String commandId, @Nullable ReadableArray args) {
        if (commandId.equals(COMMAND_REQUEST_AD)) {
            bannerView.requestAd();
        }
    }
}
